File: bindings/dart/lib/src/bgp.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// BGP protocol types for proven-servers.

/// BgpState matching the Idris2 ABI tags.
enum BgpState {
  idle(0),
  connect(1),
  active(2),
  openSent(3),
  openConfirm(4),
  established(5);

  const BgpState(this.tag);
  final int tag;

  static BgpState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// BgpEvent matching the Idris2 ABI tags.
enum BgpEvent {
  manualStart(0),
  manualStop(1),
  automaticStart(2),
  connectRetryTimerExpires(3),
  holdTimerExpires(4),
  keepaliveTimerExpires(5),
  delayOpenTimerExpires(6),
  tcpConnectionValid(7),
  tcpCrAcked(8),
  tcpConnectionConfirmed(9),
  tcpConnectionFails(10),
  bgpOpenReceived(11),
  bgpHeaderErr(12),
  bgpOpenMsgErr(13),
  notifMsgVerErr(14),
  notifMsg(15),
  keepaliveMsg(16),
  updateMsg(17),
  updateMsgErr(18);

  const BgpEvent(this.tag);
  final int tag;

  static BgpEvent? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// MessageType matching the Idris2 ABI tags.
enum MessageType {
  open(0),
  update(1),
  notification(2),
  keepalive(3);

  const MessageType(this.tag);
  final int tag;

  static MessageType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ErrorCode matching the Idris2 ABI tags.
enum ErrorCode {
  messageHeaderError(0),
  openMessageError(1),
  updateMessageError(2),
  holdTimerExpired(3),
  fsmError(4),
  cease(5);

  const ErrorCode(this.tag);
  final int tag;

  static ErrorCode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// Origin matching the Idris2 ABI tags.
enum Origin {
  igp(0),
  egp(1),
  incomplete(2);

  const Origin(this.tag);
  final int tag;

  static Origin? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// AsPathSegmentType matching the Idris2 ABI tags.
enum AsPathSegmentType {
  asSet(0),
  asSequence(1);

  const AsPathSegmentType(this.tag);
  final int tag;

  static AsPathSegmentType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// PathAttrType matching the Idris2 ABI tags.
enum PathAttrType {
  origin(0),
  asPath(1),
  nextHop(2),
  med(3),
  localPref(4),
  atomicAggr(5),
  aggregator(6),
  unknown(7);

  const PathAttrType(this.tag);
  final int tag;

  static PathAttrType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/bgp.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// BGP protocol types for proven-servers.

/** BgpState matching the Idris2 ABI tags. */
export const BgpState = Object.freeze({
  IDLE: 0,
  CONNECT: 1,
  ACTIVE: 2,
  OPEN_SENT: 3,
  OPEN_CONFIRM: 4,
  ESTABLISHED: 5,
});

/** BgpEvent matching the Idris2 ABI tags. */
export const BgpEvent = Object.freeze({
  MANUAL_START: 0,
  MANUAL_STOP: 1,
  AUTOMATIC_START: 2,
  CONNECT_RETRY_TIMER_EXPIRES: 3,
  HOLD_TIMER_EXPIRES: 4,
  KEEPALIVE_TIMER_EXPIRES: 5,
  DELAY_OPEN_TIMER_EXPIRES: 6,
  TCP_CONNECTION_VALID: 7,
  TCP_CR_ACKED: 8,
  TCP_CONNECTION_CONFIRMED: 9,
  TCP_CONNECTION_FAILS: 10,
  BGP_OPEN_RECEIVED: 11,
  BGP_HEADER_ERR: 12,
  BGP_OPEN_MSG_ERR: 13,
  NOTIF_MSG_VER_ERR: 14,
  NOTIF_MSG: 15,
  KEEPALIVE_MSG: 16,
  UPDATE_MSG: 17,
  UPDATE_MSG_ERR: 18,
});

/** MessageType matching the Idris2 ABI tags. */
export const MessageType = Object.freeze({
  OPEN: 0,
  UPDATE: 1,
  NOTIFICATION: 2,
  KEEPALIVE: 3,
});

/** ErrorCode matching the Idris2 ABI tags. */
export const ErrorCode = Object.freeze({
  MESSAGE_HEADER_ERROR: 0,
  OPEN_MESSAGE_ERROR: 1,
  UPDATE_MESSAGE_ERROR: 2,
  HOLD_TIMER_EXPIRED: 3,
  FSM_ERROR: 4,
  CEASE: 5,
});

/** Origin matching the Idris2 ABI tags. */
export const Origin = Object.freeze({
  IGP: 0,
  EGP: 1,
  INCOMPLETE: 2,
});

/** AsPathSegmentType matching the Idris2 ABI tags. */
export const AsPathSegmentType = Object.freeze({
  AS_SET: 0,
  AS_SEQUENCE: 1,
});

/** PathAttrType matching the Idris2 ABI tags. */
export const PathAttrType = Object.freeze({
  ORIGIN: 0,
  AS_PATH: 1,
  NEXT_HOP: 2,
  MED: 3,
  LOCAL_PREF: 4,
  ATOMIC_AGGR: 5,
  AGGREGATOR: 6,
  UNKNOWN: 7,
});
